src/WCSimAmBeGen.cc

```diff
@@ -0,0 +1,127 @@
+#include "WCSimAmBeGen.hh"
+#include "WCSimPrimaryGeneratorMessenger.hh"
+#include "WCSimPrimaryGeneratorAction.hh"
+
+#include "G4ParticleGun.hh"
+#include "G4PhysicalConstants.hh"
+#include "G4RandomDirection.hh"
+#include "G4SystemOfUnits.hh"
+#include "G4SPSEneDistribution.hh"
+#include "G4SPSRandomGenerator.hh"
+#include "G4ParticleTable.hh"
+#include "G4Neutron.hh"
+#include "G4Gamma.hh"
+
+#include <random>
+
+using namespace std;
+
+G4int    WCSimAmBeGen::nGammaOutcomes = 3;
+G4double WCSimAmBeGen::gammaProbabilities[3] = {0.26, 0.65, 0.08};
+G4double WCSimAmBeGen::gammaEnergies[3] = {0.0, 4.4, 7.7};
+G4int    WCSimAmBeGen::pdgids[2] = {2112, 22};
+
+WCSimAmBeGen::WCSimAmBeGen(){
+  // Initialise
+  wcsimdir = string(getenv("WCSIMDIR"))+"data/";
+
+  gs_path = wcsimdir + "ground_state_spectrum.txt";
+  fe_path = wcsimdir + "first_excited_spectrum.txt";
+  se_path = wcsimdir + "second_excited_spectrum.txt";
+
+  this->Initialise();
+}
+
+WCSimAmBeGen::~WCSimAmBeGen(){
+  
+  // This needed to be deleted
+  delete myAmBeGun;
+  delete rGen;
+  delete nEnergyDist;
+}
+
+void WCSimAmBeGen::Initialise(){
+    nEnergyDist = new G4SPSEneDistribution();
+
+    vtx = G4ThreeVector(0., 0., 0.);
+    time = 0.;
+    epsilon = 1e-6;
+}
+
+G4double WCSimAmBeGen::GammaEnergy(){
+    
+    // Initialise Geant4 random number generator
+    G4Random::setTheEngine(new CLHEP::RanecuEngine);
+
+    // Generation of the random number
+    G4double random_number = G4UniformRand();
+
+    // Select the energy taking into account the probabilities
+    G4double cumulative_probability = 0.0;
+    G4int selected_energy_index = -1;
+
+    for (G4int j = 0; j < nGammaOutcomes; ++j) {
+        cumulative_probability += gammaProbabilities[j];
+        if (random_number <= cumulative_probability) {
+            selected_energy_index = j;
+            break;
+        }
+    }
+
+    gEnergy = gammaEnergies[selected_energy_index];
+
+    return gEnergy; 
+}
+
+G4double WCSimAmBeGen::NeutronEnergy(G4double gamEnergy){
```

Comment:
This argument is coming up as an unused variable warning at compilation

I believe you have two gamma energies defined at this point
- `gamEnergy` the argument
- `gEnergy` the class variable

What I *think* you need to use is the class variable, and not have an argument to this function. Why?
- In `GenerateNG()` you create two particles
  - Neutron first, using a call to `GammaEnergy()`
  - Gamma second, using a call to `GammaEnergy()`
  - Essentially this means that the call to neutron energy is associated with a different photon energy than the actual photon energy simulated in the event

shall continue this comment elsewhere...

Reply:
Changed 👍